test/Feature/SingularSpec.hs

```diff
@@ -70,8 +70,7 @@ spec =
           `shouldRespondWith`
             [json|{"details":"Results contain 4 rows, application/vnd.pgrst.object+json requires 1 row","message":"JSON object requested, multiple (or no) rows returned"}|]
             { matchStatus  = 406
-            , matchHeaders = [ matchContentTypeSingular
-                             , "Preference-Applied" <:> "tx=commit" ]
+            , matchHeaders = [ matchContentTypeSingular ]
```

Comment:
With the refactored way we handling errors, errors go straight through `optionalRollback` untouched and this header is not being added. I believe this is fine, as the transaction is always rolled back on error and it doesn't matter what the `tx` preference was. As it's clear what happened to the user, the header is not needed according to https://tools.ietf.org/html/rfc7240#page-7 section 3.

Reply:
@wolfgangwalther WDYT?

Reply:
Hm. I think it's possible to commit a transaction and still return a `4xx` / `5xx` error code via GUC: https://postgrest.org/en/latest/api.html#setting-response-status-code

This means, that you can not tell from the error code alone whether the transaction was comitted or rolled back.

Is it possible to refactor to avoid this change?

Reply:
Thinking a bit about this @wolfgangwalther : It would be relatively simple to always set the preference header (catch any errors, turn them into a response, set the header), but I wonder whether the new behavior that emerged here is more correct.

Setting a custom status via GUC still works and will set the preference header as before, regardless of what the status code is. 

Setting the preference header will only be skipped when a true error happened and the transaction will definitely be rolled back. In those cases, showing tx=commit seems misleading.

Reply:
> Setting a custom status via GUC still works and will set the preference header as before, regardless of what the status code is.
> 
> Setting the preference header will only be skipped when a true error happened and the transaction will definitely be rolled back. In those cases, showing tx=commit seems misleading.

The whole thing is a bit more complicated, because we have `db-tx-end=commit|commit-allow-override|rollback-allow-override|rollback` settings. This creates a lot of different scenarios.

I understand RFC 7240 as follows: We should only ever return a preference that was asked for with `Preference-Applied`. This means we should never return a `Preference-Applied: tx=rollback` header, when we had a `Prefer: tx=commit` header before.

This basically means that we can not use the `Preference-Applied` header for feedback about the actual result ("committed" or "rolled back") of the transaction. A couple of examples, where this fails:

Assume an error response with `4xx` status code. With `db-tx-end=commit-allow-override` and `Prefer: tx=commit`: Setting a `Preference-Applied: tx=rollback` seems wrong. How would we differ between the "error via RAISE" and "error via GUC" cases then? We could only return nothing for RAISE (i.e. a true rollback) and `Preference-Applied: tx=commit` for the GUC case (i.e. a true commit). This basically assumes that the transaction was rolled back for any `4xx` error code by default. But now assume we have `db-tx-end=commit` and `Prefer: tx=rollback`. For a GUC error response this results in a commit in the end, because the override is not allowed. But we can't return `Preference-Applied: tx=commit` - this was not asked for. And we can't assume a rollback by default either, because that wouldn't be true.

In short: We can only use the `Preference-Applied` header to indicate what our transaction termination command was (`COMMIT` or `ROLLBACK`) - completely independent of whether the transaction was implicitly rolled back before because of an error. Note, that this is also what is implied by the config option `db-tx-end`.

There are a couple of other reasons why it makes a lot of sense to return a `Preference-Applied` header whenever we understand and accept that preference - regardless of whether it might be sensible to assume this behaviour by default in that specific case. See #740.

The current behaviour should be kept. Our goal should be to respond with more `Preference-Applied` headers instead of less.